db/system_distributed_keyspace.cc

```diff
@@ -330,6 +377,122 @@ system_distributed_keyspace::read_cdc_topology_description(
     });
 }
 
+static future<utils::chunked_vector<mutation>> get_cdc_generation_mutations(
+        const database& db,
+        utils::UUID id,
+        size_t num_replicas,
+        size_t concurrency,
+        const cdc::topology_description& desc) {
+    auto s = db.find_schema(system_distributed_keyspace::NAME_V2, system_distributed_keyspace::CDC_GENERATIONS_V2);
+
+    // To insert the data quickly and efficiently we send it in batches of multiple rows
+    // (each batch represented by a single mutation). We also send multiple such batches concurrently.
+    // However, we need to limit the memory consumption of the operation.
+    // I assume that the memory consumption grows linearly with the number of replicas
+    // (we send to all replicas ``at the same time''), with the batch size (the data must
+    // be copied for each replica?) and with concurrency. These assumptions may be too conservative
+    // but that won't hurt in a significant way (it may hurt the efficiency of the operation a little).
+    // Thus, if we want to limit the memory consumption to L, it should be true that
+    // mutation_size * num_replicas * concurrency <= L, hence
+    // mutation_size <= L / (num_replicas * concurrency).
+    // For example, say L = 10MB, concurrency = 10, num_replicas = 100; we get
+    // mutation_size <= 10MB / 1000 = 10KB.
+    // On the other hand we must have mutation_size >= size of a single row,
+    // so we will use mutation_size <= max(size of single row, L/(num_replicas*concurrency)).
+
+    // It has been tested that sending 1MB batches to 3 replicas with concurrency 20 works OK,
+    // which would correspond to L ~= 60MB. Hence that's the limit we use here.
+    const size_t L = 60 * 1000 * 1000;
```

Comment:
`60'000'000`

Reply:
It is sad that high-level code has to resort to these calculations. We should eventually move them towards the core - this code should only provide a stream of small mutations, and the core should take care of batching. But we can do this when we have more users (and so know more about the requirements).